cookie-manager.js

```diff
@@ -0,0 +1,389 @@
+        }
+
+        console.debug(`Finishing processing all cookies.`);
+    };
+
+    const getUserPreferences = function(cm_cookie) {
+
+        const cookie = getCookie(cm_cookie);
+
+        if (!cookie) {
+            return false;
+        }
+
+        try {
+            return JSON.parse(cookie);
+        } catch (e) {
+            console.error(`Unable to parse user preference cookie "${cm_cookie}" as JSON.`, e);
+            return false;
+        }
+    };
+
+    const getCookieCategoryFromManifest = function(cookie_name, configuration) {
+
+        const cookie_manifest = configuration['cookie-manifest'];
+
+        for (var i = 0; i < cookie_manifest.length; i++) {
+            const category_cookies = cookie_manifest[i]['cookies'];
+            for (var x = 0; x < category_cookies.length; x++) {
+                const cookie_prefix = category_cookies[x];
+                if (cookie_name.startsWith(cookie_prefix)) {
+                    console.debug(`Cookie "${cookie_name}" found in manifest.`);
+                    return cookie_manifest[i];
+                }
+            }
+        }
+        console.debug(`Cookie "${cookie_name}" NOT found in manifest.`);
+        return false;
+    };
+
+    const getCookie = function(cookie_name) {
+        const name = cookie_name + "=";
+        const decoded_cookie = decodeURIComponent(document.cookie);
+        const cookie_array = decoded_cookie.split(';');
+
+        for(var i = 0; i < cookie_array.length; i++) {
+            let cookie_part = cookie_array[i];
+            while (cookie_part.charAt(0) === ' ') {
+                cookie_part = cookie_part.substring(1);
+            }
+            if (cookie_part.indexOf(name) === 0) {
+                return cookie_part.substring(name.length, cookie_part.length);
+            }
+        }
+        return false;
+    };
+
+    const deleteCookie = function(cookie_name) {
+        document.cookie = cookie_name + '=; expires=Thu, 01 Jan 1970 00:00:01 GMT; path=/;';
+        document.cookie = cookie_name + '=; expires=Thu, 01 Jan 1970 00:00:01 GMT;domain='+window.location.hostname +';path=/;';
+        document.cookie = cookie_name + '=; expires=Thu, 01 Jan 1970 00:00:01 GMT;domain=.'+window.location.hostname +';path=/;';
+
+        let firstDot = window.location.hostname.indexOf('.');
+        let upperDomain = window.location.hostname.substring(firstDot);
+        document.cookie = cookie_name + '=; expires=Thu, 01 Jan 1970 00:00:01 GMT;domain='+upperDomain+';path=/;';
+        console.debug(window.location.hostname);
+        console.debug(upperDomain);
+
+        console.debug(`Deleted cookie "${cookie_name}"`);
+    };
+
+    const setCookie = function(configuration, cookie_value) {
+
+        const cookie_name = configuration['user-preference-cookie-name'];
+
+        let cookie_secure = configOptionIsTrue(configuration, 'user-preference-cookie-secure');
```

Comment:
Suggest we need some soft of style checking tool for the next release, I can see a few inconsistencies, but ok for this iteration.

e.g. `let cookie_secure` vs `let firstDot` or spaces after an `if`